### ISSUE TITLE
lib: (re-)export all `builtins` functions

### ISSUE THREAD

**Issue by stepbrobd** [NixOS/nixpkgs#369215]
I remember couple months ago in The Nix Hour #76 ([specifically around this timestamp](https://youtu.be/G_KGcVjaS5c?t=241)), @infinisil mentioned ideally nixpkgs should export all builtin symbols, and I very very strongly agree with this.

Could we possibly expose all symbols by doing something like this (not exactly like this, but internally somewhere in `lib/default.nix`):

```nix
lib = with lib; extend (_: _: (removeAttrs builtins (intersectLists (attrNames lib) (attrNames builtins))))
```

This way, functions only exist in `builtins` can be re-exported, while not replacing functions with different implementations in `nixpkgs.lib` (like `functionArgs`), and potentially, new `nixpkgs.lib` functions.

Also there's the question whether this proposal is in conflict with [this TODO](https://github.com/NixOS/nixpkgs/blob/082903a098e64f44e26fe12d75c1e34073e4456d/lib/default.nix#L70).

**Comment by infinisil** [comment]
We shouldn't blindly forward all builtins, especially because builtins change depending on the Nix version, Nixpkgs needs to be independent of Nix. But also because I'm not sure if all builtins would fit into `lib`, here's a quick brainstorm of which I don't think would fit:
- `langVersion`, `nixVersion`, `nixPath`, `currentSystem`, `import`, `scopedImport`, `storeDir` (impure)
- `abort`/`throw`, `null`/`true`/`false`, `builtins` (effectively language features, accessible without `builtins.`)
- `toPath` (deprecated)
- `derivation`/`derivationStrict`, `fetchGit`/`fetchTarball`/... (complex behavior, probably doesn't fit into `lib`, confusable with `pkgs.fetch*`)
- `findFile` (bad name)
- (maybe others)

**Comment by stepbrobd** [comment]
I agree that the use of some `builtins` should be discouraged (as listed above), but we should carefully choose where to enforce those rules.

IMO there are three sides of `nixpkgs.lib`:

- within nixpkgs:
  - (1) what `lib` maintainers are allowed to use, i.e. usages within the `lib/` directory
  - (2) what maintainers can use inside the nixpkgs repo but external to `lib/` directory (like in `maintainers/`, `nixos/`, `pkgs/`, etc.)
- externally (when using nixpkgs as a dependency)
  - (3) exported functions users can call

For (1), I personally don't see any issue with using impure functions and we already have quite a few of them in there

For (2), I favor we export all **allowed** `builtins` along with the standard `nixpkgs.lib` implementation. Module/package maintainers should only use exported functions (and their own custom functions, but we strictly disallow the use of `builtins`).

For (3), no restriction, export full `builtins` + `nixpkgs.lib`, and standard `nixpkgs.lib` implementation take precedence when there's a name collision)

**Comment by Atemu** [comment]
If we're going to re-export all builtins, might as well simplify to `builtins // lib`.

I don't really see a point in trying to hide impure builtins. If users don't want impure functions, they should enable pure eval and will get an eval error if they try to use any.

**Comment by overby-me** [comment]
@eclairevoyant are all builtins exported now?

**Comment by Malix-Labs** [comment]
https://github.com/NixOS/nixpkgs/issues/369215#issuecomment-4841605408 from @overby-me :

> are all builtins exported now?

```console
❯ nix eval --json --impure --expr 'let lib = import ./lib; in with builtins; with lib; subtractLists (attrNames lib) (attrNames builtins)'
[
  "abort",
  "break",
  "builtins",
  "currentSystem",
  "currentTime",
  "derivation",
  "derivationStrict",
  "false",
  "fetchGit",
  "fetchMercurial",
  "fetchTarball",
  "fetchTree",
  "fetchurl",
  "findFile",
  "getEnv",
  "getFlake",
  "import",
  "isNull",
  "langVersion",
  "nixPath",
  "nixVersion",
  "null",
  "scopedImport",
  "storeDir",
  "storePath",
  "throw",
  "toPath",
  "traceVerbose",
```

No

This issue should be reopened

**Comment by Malix-Labs** [comment]
I wonder if a test can be made for that to be caught automatically

**Comment by Atemu** [comment]
I'd still like to know why we don't just simply `lib = builtins // { /* lib definitions */ }`.

If we wanted to hide some builtins (why?), I'd prefer a blocklist (`removeAttrs [ ... ] builtins`) rather than an allowlist.

**Comment by Malix-Labs** [comment]
@Atemu please see https://github.com/NixOS/nixpkgs/pull/541081#issuecomment-4961780447

**Comment by Malix-Labs** [comment]
**Requesting reviews** for which should be re-exported and which should be implemented here : https://github.com/NixOS/nixpkgs/pull/541081#discussion_r3575375418
